toolchain/check/dump_id.cpp

```diff
@@ -0,0 +1,50 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/check/dump_id.h"
+
+#include "common/check.h"
+#include "llvm/Support/raw_ostream.h"
+#include "toolchain/check/context.h"
+#include "toolchain/lex/tokenized_buffer.h"
+#include "toolchain/parse/tree.h"
+#include "toolchain/sem_ir/file.h"
+
+namespace Carbon::Check {
+
+auto DumpIdImpl(SemIR::LocId loc_id, const Context& context) -> void {
+  if (!loc_id.is_valid()) {
+    llvm::errs() << "LocId(invalid)";
+    return;
+  }
+
+  if (loc_id.is_node_id()) {
+    auto token = context.parse_tree().node_token(loc_id.node_id());
+    auto line = context.tokens().GetLineNumber(token);
+    auto col = context.tokens().GetColumnNumber(token);
+    const char* implicit = loc_id.is_implicit() ? " implicit" : "";
+    llvm::errs() << "LocId(";
+    llvm::errs().write_escaped(context.sem_ir().filename());
+    llvm::errs() << ":" << line << ":" << col << implicit << ")";
+  } else {
+    CARBON_CHECK(loc_id.is_import_ir_inst_id());
+
+    auto import_ir_id = context.sem_ir()
+                            .import_ir_insts()
+                            .Get(loc_id.import_ir_inst_id())
+                            .ir_id;
+    const auto* import_file =
+        context.sem_ir().import_irs().Get(import_ir_id).sem_ir;
+    llvm::errs() << "LocId(import from \"";
+    llvm::errs().write_escaped(import_file->filename());
+    llvm::errs() << "\")";
+  }
+}
+
+template <>
+auto DumpIdMethods<Context>::Newline() const -> void {
+  llvm::errs() << "\n";
+}
```

Comment:
Why the function for this instead of doing the write directly? Maybe the function should have some comments if it's needed?

Also, maybe this should be `WriteNewline`? We typically name functions describing the action being taken, and it actually was unclear to me what `Newline` would do.

Reply:
I wanted to avoid putting streaming includes into the header file, since operator<< is bad for compile times. Is this something we think about here? I took a quick look around and saw that the streaming includes had been generally avoided in headers elsewhere so far.

Reply:
I'm not sure which include you're looking for, but "common/ostream.h" seems to be relatively common.

Reply:
Ok noted, thanks, I will not try to avoid specific includes in header files then. I moved to using this header everywhere.